### ISSUE TITLE
Is node label a necessary?

### ISSUE THREAD

**Issue by YamamotoSayaka** [lucfra/LDS-GNN#1]
Hi, thanks for this nice implementation and paper.
From my understanding,  LDS-GNN updates graphs and GCN parameters at the same time for a node classification problem. So, say if in another task, the object is to find the label of the edge ( while there's no label for vertices), can LDS-GNN still fit into this task?

Many thanks.

**Comment by lucfra** [comment]
Hi Yamamoto,

Thanks for your post and for your interest in our work!
So... the short answer is that I believe you can try to modify the code to do that, provided you can specify an inner and outer object that you wish to optimize.
But I'm not totally sure I understand what you have in mind :)  
Specifically, are you thinking about a supervised learning task? In this case which kind of supervision are you assuming to have? Are you assuming that you have an incomplete graph and a training set with some labelled edges or are you more generally thinking about multi-relational graphs?

Cheers, 
Luca

**Comment by YamamotoSayaka** [comment]
Hi @lucfra 
From equation (5) from the original paper, we are trying to find the optimal adj matrix A and graph parameters at the same time, when the labels of the vertices are already known, right?

However, my problem is that, given a training set with only labelled edges, so for example my graph has 5 vertices and only two of them are connected (but no vertices labels are provided), and I want to find the which edges are connected given another graph. Does LDS-GNN still fit?

Ahhhh I'm quite new to graph learning, maybe this problem is somewhat naive.
Many thanks any ways!

**Comment by lucfra** [comment]
Hi Yamamoto,

yes. You're correct. We try to find the optimal adjacency matrix A (well, specifically that that minimizes the error on a validation set of nodes) and jointly the parameters of a GCN in a so-called transductive (semi-supervised) learning setting. 
What we are ultimately interested in is to find a good model that predicts the labels of the test nodes (nodes for which we only have the input features and, possibly -- but not necessarily -- an (incomplete) list of edges). To achieve this, we propose to work on the relationship structure among datapoints, i.e. the adjacency matrix. 

At the moment, the work does not straightforward generalize to a setting with more than one graph, as you should for example need a much more powerful generative model for the graph itself (we used independent Bernoulli random vairables) and to reformulate the problem accordingly.

I hope this helps; but I'm not totally sure I've fully understood the setting that you have in mind. Specifically, what do you mean exactly by `labelled edges`? Does each edge of the graph have a class?

Cheers, 
Luca

**Comment by YamamotoSayaka** [comment]
@lucfra Thx! I think I get it

**Comment by brandhsu** [comment]
Can this methodology also learn the relations (adjacency matrix) between nodes given the task of graph classification assuming one is given a set of nodes without an adjacency matrix that is assigned to a corresponding class?

**Comment by lucfra** [comment]
Hi @Brandhsu ,

We developed and presented LDS as an algorithm for semi-supervised (node) classification problems. LDS can be easily adapted to (node) regression problems, but not to graph classification problems. This is because LDS learns an adjacency matrix in a transductive setting. 

The way that I see to tackle the problem you have probably in mind is to devise an inductive strategy to learn the adjacency matrices, but this surely requires a bit of work.

Hope this helps.

Cheers, 
Luca

**Comment by zeyuyun1** [comment]
Hi @lucfra 

Could LDS used to induce graphs for a standard classification problem?

For example, my datasets consist of data point and label (x,y). I want to perform the standard classification problem y = f(x). The data x looks like x = [x_1,x_2,x_3,...], where each x_i is in some high dimension. There's an underlying graph structure within x, that is also invaraint for all x. What I can do is to create a simple KNN graph on x, and cast this to a whole graph classification problem. But this might not be the most idea, that's why I wonder if I can use LDS to induce a better graph. 

Thank you!
Zeyu

**Comment by lucfra** [comment]
Hi @zeyuyun1 

Thanks for your message and interest in our work. 
If I understand correctly, what you say makes sense (and LDS could help!) as long as you also have your test points available. I'm not sure about what you mean by ``There's an underlying graph structure within x, that is also invariant for all x. '' though

LDS is a transductive approach: this means that graphs are not induced from data, but are rather learnt for a particular set of data (and are not directly ``transferable''). 

Note also that if you don't give a starting graph LDS initialize the graph precisely with kNN, and then optimizes it.

Hope this helps, 
Cheers,
Luca

**Comment by zeyuyun1** [comment]
Hi @lucfra,

Thank you for the quick reply. Sorry for the confusion. Let me clarify the part ``There's an underlying graph structure within x, that is also invariant for all x. ''

In a classical graph classification task, each data point is (x, a), where x is node features with dimension (num of nodes, num of features), and a is a graph with dimension (num of nodes, num of nodes). The entire dataset would be (X, A). For example, in molecule structure classification, a is the structure of a molecule and x represents the atom at each position. This is a graph-level classification task because we're predicting what type of molecule (label) for each given node features and graph.

My dataset is the same setup as above, except all the graph in the dataset A is the same. Imagine a traffic system of 100 intersections. There's only one graph, the one that describes the connectivity of those intersections. But given the states of each traffic light (node features), we want to predict whether or not the city is in rush hour.

Moreover, even if the graph of the traffic system is not provided, but only the position of each intersection is provided, if we're able to implement KNN-LDS on this dataset, we could potentially learn the traffic system itself, or an "optimal" traffic system.

If I understand correctly, the reason why the molecule classification task doesn't work is that your work only parametrizes the statistic of one graph. If we want to account for the case where there are multiple possible graphs, we need some sort of a generator network. However, you mentioned this at the end of your paper:

"While relatively efficient, it cannot currently scale to large datasets: this would require an implementation that works with mini-batches of nodes. "

In the dataset I described above, even though I only have one graph, I still probably need to use some stochastic gradient descent since the dataset is large. Would this be a problem?

Sorry for the long description. Thank you so much for your help!

**Comment by lucfra** [comment]
Hi @zeyuyun1 

now I see. Thanks for the clarification; this also sounds like a pretty interesting application!
Yes, in the case you mention, I'd say LDS can be used more or less as is (you just need to redefine the training and validation errors), provided that all fits in memory. 

For scaling up LDS for your problem, it might be sufficient to implement stochasticity along the axis of the various examples (the x's), keeping the adjacency graph between iterations. This shouldn't be too difficult to implement. 
The more challenging part is if you have to implement stochasticity within the graph, i.e. if single instances are too big, and you need to subsample at each iteration a subset of nodes and subsets of their neighbourhoods. 

If anything is not clear enough, feel free to write me back!

Cheers, 
Luca

PS: On a slightly different note, we've just published on ArXiv a paper that might interest you: https://arxiv.org/abs/2106.01798 . It's not exactly a follow-up of LDS, but it's quite related.

**Comment by zeyuyun1** [comment]
Hi @lucfra 

Thanks so much for your help. I spent last week trying to figure this out... I think the only thing I don't get now is how to implement the stop condition. More specifically:

In your implementation, all datapoints are available during each iteration, so you can just report the training accuracy at the current itertation. However, if I have stochasticity along the axis of the various examples, then we need to accumulate a bunch of training points, then report the training accuracy for the stop condition. 

I think the pseudo-code below explains pretty clearly what I want to do, I changed [335-337](https://github.com/lucfra/LDS-GNN/blob/ba580f9b19fff9329f089b289d52929623be3cb2/lds_gnn/lds.py#L335-L337) to the following:

            for data,label in data_loader:  # inner optimization loop
                tr_fd={
                        X: data,
                        Y_in: label,
                        keep_prob:0.5
                    }
                ho_step(1, tr_fd, tr_fd, online=j)  # do one hypergradient optimization step

In this case, it seems like I just didn't use the following line, and the performance is not good.

                steps, rs = ho_mod.hypergradient.min_decrease_condition(
                    config.io_params[0], config.io_params[1], config.io_steps,
                    feed_dicts=tr_fd, verbose=False, obj=None)

Any help will be appericated! Again, thank you so much for the help so far.

**Comment by lucfra** [comment]
Hi @zeyuyun1 , 

I'm not sure I understand if you're referring to the stopping condition for the inner or outer problem.
Now that I think of it, how many GNNs (weights) do you have currently/would you like/is fine to have?
Because I can imagine at least 2 settings whereby in (A) you only have one single GNN and in (B) you have one GNN per example (i.e. a set of points). Setting (B) is basically a meta-learning setting, not sure if this would be applicable in your case...

Anyway, in setting (B) I'd say that for the inner, as you're processing 1 graph at a time (or a mini-batch?) you can still check the decrease condition of the inner objective for that particular graph (and that one only!).
For the outer condition, I'd probably hold out some graphs and do early stopping by computing accuracy on those at regular interval. This probably would require running the inner algorithm first, i.e. optimizing the weights of a GNN with the points of those graphs (but without changing the adjacency matrix).

In setting (A) it might be more tricky to get it right for the inner condition; you might try simply setting a number of epochs N and treat N as a hyperparameter. I'd say this is the safest course of action at first..

To be honest with you, I do not remember the details of the implementation of `ho_mod.hypergradient.min_decrease_condition` . 
Did you change its usage?

Hope this helps, 
Cheers, 
Luca

**Comment by zeyuyun1** [comment]
Hi @lucfra ,

I am thinking about the stopping condition for the inner problem.

I think my setting is (B). In general, whole graph classification should be in (B) setting. For nodes classification (what's in the paper), each nodes has an label, so it would make sense to fit a GNN to the entire graph. However, For whole graph classification, each example (i.e. a set of nodes) only have one label. In this case, trying to fit a GNN to each example wouldn't really make sense. 

I didn't change the usage of ho_mod.hypergradient.min_decrease_condition,  I think it basically just check if the validation accuracy still increase in the current iteration (it has some tolerance: 'dec' and 'patient' parameter) for inner objective.

I am not sure if I understood 
`In setting (A) it might be more tricky to get it right for the inner condition; you might try simply setting a number of epochs N and treat N as a hyperparameter.`

Let me be more specific on the datasets. Say dataset has 20,000 examples (20,000 sets of nodes vector), each nodes vector represents the state of the current location (e.g. traffic light signal, numbers of cars at certain time). Each example then correspond to a binary label "0:  there's traffic 1: no traffic." And we're trying to get one optimal graph, or one parameter "theta" for all those examples.

In this case, it wouldn't make sense to calculate the validation accuracy on a single example, because it will just result in 0 or 1. It would only make sense if we run on 2,000 examples, then calculate the accuracy. In this case, what do you mean by "try simply setting a number of epochs N"?

Thank you!
Zeyu

**Comment by lucfra** [comment]
Right, I see. 

Then you're probably in something close to setting A, i.e. you have 1 GNN and fit the GNN weights across the 20000 examples, isn't it?

What I meant is that you can try to remove the stopping condition altogether, and simply do a certain predefined number of steps/epochs E to minimize your inner objective:

Another possible option (which I would try after the first) would be to implement something like early stopping at the inner level: hold out a portion of examples, say 2000, and periodically check whether your accuracy score keeps on decreasing there. 

Hope this is clearer.

Cheers, 
Luca
